### PR TITLE
Management command to clear owner IDs on cases of a specific case type

### PR DESCRIPTION
**Pull request by AddisonDunn**
## Product Description

This management command that can be run by a developer will clear out the owner ID field for all cases of a given case type inside a given domain. 

The USH CI/CT Alaska project has requested for the owner IDs of certain case types inside their production domain to be cleared (that is, set to a dash "-"). This cannot be done properly via case importer and has a deadline of the end of the year. See the ticket linked below.

## Technical Summary

[USH-1469](https://dimagi-dev.atlassian.net/browse/USH-1469). 

Example command execution:
`./manage.py clear_owner_ids addison-test person`

This management command was added as a series of USH project-related commands in `custom/covid/management/commands` that use a base class `CaseUpdateCommand`.

The command is relatively simple and inherits much of its syntax and methods from the other Covid management commands. One item of note, I almost chose to use the `update_case` method from `corehq.apps.hqcase.utils` instead of creating a series of case blocks as many of the other `custom/covid` commands do. It was much simpler, but I thought it was more important to keep the code standardized in this part of the codebase and to specify the same device ID as other commands and keep the option open to specify the user ID (as the other Covid commands do).

This will require a private release to run this management command before 12/31/2021.

### Safety story

I tested locally to make sure the command runs as expected and doesn't affect other case types. I'll be testing on staging before we privately release this.

### Automated test coverage

~~None~~ 
Test added.

### QA Plan

None, as far as I know management commands are outside of QA's scope.

### Migrations

No migrations.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by esoergel** [comment]
This test failed 5 times.  It's marked as flaky and I don't see how it could be related to this PR, so I went ahead and restarted the build.
https://github.com/dimagi/commcare-hq/blob/f0e5d829b75315188ce707ad299b16f0f0b32fc9/corehq/apps/sms/tests/test_backends.py#L922-L923

File: custom/covid/management/commands/clear_owner_ids.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+BATCH_SIZE = 100
+DEVICE_ID = __name__ + ".update_owner_ids"
+
+
+class Command(CaseUpdateCommand):
+    help = "Makes the owner_id for cases blank"
+
+    def case_block(self, case):
+        blank_owner_id = {'owner_id': '-'}
+        return ElementTree.tostring(CaseBlock.deprecated_init(
```
Comment (esoergel) [inline custom/covid/management/commands/clear_owner_ids.py:19]:
```suggestion
        return ElementTree.tostring(CaseBlock(
```

For context:
https://github.com/dimagi/commcare-hq/blob/f0e5d829b75315188ce707ad299b16f0f0b32fc9/corehq/ex-submodules/casexml/apps/case/mock/case_block.py#L61-L69

File: custom/covid/management/commands/clear_owner_ids.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    def update_cases(self, domain, case_type, user_id):
+        case_ids = self.find_case_ids_by_type(domain, case_type)
+        accessor = CaseAccessors(domain)
+        case_blocks = []
+        errors = []
+
+        print(f"Found {len(case_ids)} {case_type} cases in {domain}")
+
+        for case in accessor.iter_cases(case_ids):
+            case_blocks.append(self.case_block(case))
```
Comment (esoergel) [inline custom/covid/management/commands/clear_owner_ids.py:34]: nit: you can skip applying updates for cases that already have the owner ID set to `-`
```suggestion
            if case.owner_id != "-":
                case_blocks.append(self.case_block(case))
```

**Review by esoergel — COMMENTED** [review]
I have some thoughts I want to note about this migration process, but as this instance is largely in line with what we've been doing without issue, they can be treated as "food for thought" type comments:

I like having manual confirmation - "you're about to change X cases in Y way, are you sure? (y/n)"

Some more detailed logging would be helpful if we need to troubleshoot - what are the IDs of the forms submitted?  What were the before/after values of the case properties affected?  We arguably get this for free with CommCare's data integrity model - you can search forms submitted with this device ID, so that might not actually be strictly necessary.

We fetch cases in chunks of 100 and save them in chunks of 100, but we do these things in sequence, which means that the list of case blocks must hold all changes on the domain in memory.  This could cause issues if that list is big enough (though this isn't the first migration that takes that approach, so I guess it doesn't).  An easy alternative approach would be to do the chunked fetching here, calling `submit_case_blocks` at the end of each chunk.  Or if we want to keep them separate (useful if we're only modifying a subset of cases and/or want to submit exactly 100 updates at a time), submit the list whenever it hits a length of 100, similar to what [`IterDB` does](https://github.com/dimagi/commcare-hq/blob/f0e5d829b75315188ce707ad299b16f0f0b32fc9/corehq/util/couch.py#L210-L213).

File: custom/covid/management/commands/clear_owner_ids.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        total_cases_updated = 0
+        for chunk in chunked(case_blocks, BATCH_SIZE):
+            submit_case_blocks(chunk, domain, device_id=DEVICE_ID, user_id=user_id)
+            total_cases_updated += len(chunk)
+            print("Updated {} cases on domain {}".format(total_cases_updated, domain))
+
+        print(f"{total_cases_updated} cases of case type {case_type} in domain {domain} had their owner ID"
+            " field cleared successfully.")
+
+        self.log_data(domain, "clear_owner_ids", case_type, len(case_ids), total_cases_updated, errors)
```
Comment (esoergel) [inline custom/covid/management/commands/clear_owner_ids.py:45]: `errors` isn't actually used - I'd just define it in line
```suggestion
        self.log_data(domain, "clear_owner_ids", case_type, len(case_ids), total_cases_updated, errors=[])
```

**Review by esoergel — COMMENTED** [review]
forgot to submit this comment

File: custom/covid/management/commands/clear_owner_ids.py (hunk 2 above)
Comment (AddisonDunn) [reply custom/covid/management/commands/clear_owner_ids.py]: Thanks, added.

File: custom/covid/management/commands/clear_owner_ids.py (hunk 1 above)
Comment (AddisonDunn) [reply custom/covid/management/commands/clear_owner_ids.py]: Nice, added.

**Comment by AddisonDunn** [comment]
@esoergel I added most of your suggestions but as we discussed I left out the logging because outputting to a CSV would be a bit of a pain. The only other point of anxiety I have is with the change to manual batching, which because I haven't gotten celery set up locally for imports I haven't been able to test on a case number greater than 100. But there are 1000+ cases in the Alaska test domain so testing on that should be sufficient.

File: custom/covid/management/commands/clear_owner_ids.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            if(case.get_case_property('owner_id') != '-'):
+                case_blocks.append(self.case_block(case))
+            else:
+                count_already_blank += 1
+            if(len(case_blocks) >= BATCH_SIZE):
+                submit_case_blocks(case_blocks, domain, device_id=DEVICE_ID, user_id=user_id)
+                case_blocks = []
+                total_cases_updated += BATCH_SIZE
+
+        if(case_blocks):
```
Comment (esoergel) [inline custom/covid/management/commands/clear_owner_ids.py:43]: Nitpick: while valid, these `if` statements are formatted weirdly
```suggestion
        if case_blocks:
```

**Review by esoergel — APPROVED** [review]

File: custom/covid/management/commands/clear_owner_ids.py (hunk 4 above)
Comment (AddisonDunn) [reply custom/covid/management/commands/clear_owner_ids.py:43]: Old Java/JS habits die hard. Will fix, thanks Ethan.

**Review by esoergel — APPROVED** [review]

**Comment by AddisonDunn** [comment]
Merging. FYI to all this was run successfully and without hiccup for the relevant ticket, and can be used down the line.

**Pull request merged**